packages/multicast_dns/lib/src/packet.dart

```diff
@@ -0,0 +1,373 @@
+        }
+        return IPAddressResourceRecord(fqdn, validUntil,
+            address: InternetAddress(addr.toString()));
+      case ResourceRecordType.aaaa:
+        checkLength(offset + rDataLength);
+        final StringBuffer addr = StringBuffer();
+        final int stop = offset + rDataLength;
+        addr.write(packetBytes.getUint16(offset).toRadixString(16));
+        offset += 2;
+        for (offset; offset < stop; offset += 2) {
+          addr.write(':');
+          addr.write(packetBytes.getUint16(offset).toRadixString(16));
+        }
+        return IPAddressResourceRecord(
+          fqdn,
+          validUntil,
+          address: InternetAddress(addr.toString()),
+        );
+      case ResourceRecordType.srv:
+        checkLength(offset + 2);
+        final int priority = packetBytes.getUint16(offset);
+        offset += 2;
+        checkLength(offset + 2);
+        final int weight = packetBytes.getUint16(offset);
+        offset += 2;
+        checkLength(offset + 2);
+        final int port = packetBytes.getUint16(offset);
+        offset += 2;
+        final _FQDNReadResult result =
+            _readFQDN(data, packetBytes, offset, length);
+        offset += result.bytesRead;
+        return SrvResourceRecord(
+          fqdn,
+          validUntil,
+          target: result.fqdn,
+          port: port,
+          priority: priority,
+          weight: weight,
+        );
+        break;
+      case ResourceRecordType.ptr:
+        checkLength(offset + rDataLength);
+        final _FQDNReadResult result =
+            _readFQDN(data, packetBytes, offset, length);
+        offset += rDataLength;
+        return PtrResourceRecord(
+          fqdn,
+          validUntil,
+          domainName: result.fqdn,
+        );
+      case ResourceRecordType.txt:
+        checkLength(offset + rDataLength);
+        final Uint8List rawText = Uint8List.view(
+          data.buffer,
+          offset,
+          rDataLength,
+        );
+        final String text = utf8.decode(rawText);
+        offset += rDataLength;
+        return TxtResourceRecord(fqdn, validUntil, text: text);
+      default:
+        checkLength(offset + rDataLength);
+        offset += rDataLength;
+        return null;
+    }
+  }
+
+  // Note: this list can't be fixed length right now because we might get
+  // resource record types we don't support, and consumers expect this list
+  // to not have null entries.
+  final List<ResourceRecord> result = <ResourceRecord>[];
+
+  try {
+    for (int i = 0; i < ancount + arcount; i++) {
+      final ResourceRecord record = readResourceRecord();
+      if (record != null) {
+        result.add(record);
+      }
+    }
+  } on MDnsDecodeException {
+    // If decoding fails return null.
+    return null;
+  }
+  return result;
+}
+
+/// Exceptions thrown by decoder when the packet is invalid.
+class MDnsDecodeException implements Exception {
+  /// Creates a new MDnsDecodeException.
```

Comment:
This comment is against the style guide: it should be more descriptive. Say what values offset can take on, for instance. Can it be null?  If not, it should probably also have an assert to that effect.